zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs

```diff
@@ -118,89 +122,70 @@ impl InputSource for TransactionRecordsById {
-                        })?)
-                }
-                None => Ok(None),
-            },
-        )? {
-            if let Some(missing_value_after_orchard) =
-                orchard_note_noteref_pairs.into_iter().try_fold(
-                    Some(missing_value_after_sapling),
-                    |rolling_target, (note, note_id)| match rolling_target {
-                        Some(targ) if targ == NonNegativeAmount::ZERO => Ok(None),
-                        Some(targ) => {
-                            orchard_notes.push(
-                                self.get(note_id.txid())
-                                    .and_then(|tr| {
-                                        tr.get_received_note::<OrchardDomain>(
-                                            note_id.output_index() as u32,
-                                        )
-                                    })
-                                    .ok_or(InputSourceError::NoteCannotBeIdentified(note_id))?,
-                            );
-                            Ok(targ
-                                - NonNegativeAmount::from_u64(note.value().inner()).map_err(
-                                    |e| InputSourceError::OutputTooBig((note.value().inner(), e)),
-                                )?)
-                        }
-                        None => Ok(None),
-                    },
-                )?
-            {
-                if missing_value_after_orchard != NonNegativeAmount::ZERO {
-                    return Err(InputSourceError::Shortfall(
-                        missing_value_after_orchard.into_u64(),
-                    ));
-                }
-            };
-        };
+        let mut selected_sapling = Vec::<ReceivedNote<NoteId, sapling_crypto::Note>>::new();
+        let mut selected_orchard = Vec::<ReceivedNote<NoteId, orchard::Note>>::new();
+
+        // transform each NoteId to a ReceivedNote
+        selected.iter().try_for_each(|(id, _value)| {
+            let transaction = self.get(id.txid());
+            let output_index = id.output_index() as u32;
+            match id.protocol() {
+                zcash_client_backend::ShieldedProtocol::Sapling => transaction
+                    .and_then(|transaction_record| {
+                        transaction_record.get_received_note::<SaplingDomain>(output_index)
+                    })
+                    .map(|received_note| {
+                        selected_sapling.push(received_note);
+                    }),
+                zcash_client_backend::ShieldedProtocol::Orchard => transaction
+                    .and_then(|transaction_record| {
+                        transaction_record.get_received_note::<OrchardDomain>(output_index)
+                    })
+                    .map(|received_note| {
+                        selected_orchard.push(received_note);
+                    }),
+            }
+            .ok_or(InputSourceError::NoteCannotBeIdentified(*id))
```

Comment:
this works but looks messy to me, could we handle each option and return an error in place to avoid using `and_then` and `map` and then returning an Option<()> ? I think it would be much more readable
```suggestion
                    // transform each NoteId to a ReceivedNote
        for (id, _value) in selected {
            let transaction = self
                .get(id.txid())
                .ok_or(InputSourceError::NoteCannotBeIdentified(id))?;
            let output_index = id.output_index() as u32;
            match id.protocol() {
                zcash_client_backend::ShieldedProtocol::Sapling => {
                    let received_note = transaction
                        .get_received_note::<SaplingDomain>(output_index)
                        .ok_or(InputSourceError::NoteCannotBeIdentified(id))?;
                    selected_sapling.push(received_note);
                }
                zcash_client_backend::ShieldedProtocol::Orchard => {
                    let received_note = transaction
                        .get_received_note::<OrchardDomain>(output_index)
                        .ok_or(InputSourceError::NoteCannotBeIdentified(id))?;
                    selected_orchard.push(received_note);
                }
            }
        }

```

Reply:
i really like the pattern with and_then. updating identifier for clarity